[docs] Improve the propTypes generation and API demos' links

mnajdova opened the pull request:
This PR adds support for System components on the docs. Changes done:
- ~generate prop types for the components inside the system~ (manually added `propTypes` for the `Box` component)
- adds Components and API sections in the system documentation
- moved the box page under the Components section (`system/box` -> `system/react-box`)
- adds support for building APIs for the system components (generated with `createX` pattern)
- the system API pages link demos to all products (with specifying the product next to the demos name)
- the Material UI API pages link demos only to the Mateiral UI components demos to avoid confusion

siriwatknp commented on docs/public/_redirects:
```diff
@@ -237,6 +237,8 @@ https://v4.material-ui.com/* https://v4.mui.com/:splat 301!
 /guides/classname-generator/ /experimental-api/classname-generator/ 301
 /:lang/guides/classname-generator/ /:lang/experimental-api/classname-generator/ 301
 
+/system/box/ /system/react-box/ 301
```
```suggestion
/system/box/ /system/react-box/ 301
/:lang/system/box/ /:lang/system/react-box/ 301
```

siriwatknp commented:
Overall, looks good to me except for the Joy's Box as you described 😅

mnajdova commented:
> Overall, looks good to me except for the Joy's Box as you described 😅

Great, I will start with that then 👍

mnajdova commented:
@siriwatknp it was a false positive that the `yarn proptypes` worked for any of the `Box` components. The problem is that the `typescript-to-proptypes` does not recognize the `Box` as a component, as it basically supports the following based on their [docs](https://github.com/merceyz/typescript-to-proptypes):

Component type |  
-- | --
Class | ✔️
Function | ✔️
Const functions | ✔️
React.memo | ✔️
React.ForwardRef | ✔️

I guess we will need to manually add the propTypes on the system components (I think this should be fine as it is a small set of components). I've updated the message above the prop types to state this. While working on this, I noticed a bug in the joy generation of proptypes that I've fixed (you can take a look at the proptypes generated to validate if we want this change. I am updating the title to better match the changes.

siriwatknp commented:
> While working on this, I noticed a bug in the joy generation of proptypes that I've fixed (you can take a look at the proptypes generated to validate if we want this change

If you meant the `sx` proptype generated in Joy, I think that's fine (Material UI components also have it).

siriwatknp reviewed (approved):
👍 Awesome. Looks good to me!
